lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -25,78 +25,240 @@
 logger = logging.getLogger(__name__)
 
 
-class OpenSearchOpsLock:
-    """This class covers the configuration changes depending on certain actions."""
-
-    LOCK_INDEX = ".ops_lock"
-    PEER_DATA_LOCK_FLAG = "ops_removing_unit"
+class _PeerRelationEndpoint(ops.Object):
+    _NAME = "node-lock-fallback"
 
-    def __init__(self, charm):
+    def __init__(self, charm: ops.CharmBase):
+        super().__init__(charm, self._NAME)
         self._charm = charm
-        self._opensearch = charm.opensearch
+        self.framework.observe(
+            self._charm.on[self._NAME].relation_changed, self._on_peer_relation_changed
+        )
+
+    @property
+    def acquired(self) -> bool:
+        """Attempt to acquire lock.
 
-    @retry(stop=stop_after_attempt(3), wait=wait_fixed(0.5), reraise=True)
-    def acquire(self):
-        """Method for Acquiring the "ops" lock."""
-        # no lock acquisition needed if only 1 unit remaining
-        if len(self._charm.model.get_relation(PeerRelationName).units) == 1:
+        Returns:
+            Whether lock was acquired
+        """
+        if not self._relation:
+            return False
+        self._relation.data[self._charm.unit]["lock-requested"] = json.dumps(True)
+        if self._charm.unit.is_leader():
+            logger.debug("[Node lock] Requested peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+        acquired = self._unit_with_lock == self._charm.unit.name
+        if acquired:
+            logger.debug("[Node lock] Acquired via peer databag")
+        else:
+            logger.debug(
+                f"[Node lock] Not acquired. Unit with peer databag lock: {self._unit_with_lock}"
+            )
+        return acquired
+
+    def release(self):
+        """Release lock for this unit."""
+        if not self._relation:
             return
+        self._relation.data[self._charm.unit].pop("lock-requested", None)
+        if self._charm.unit.is_leader():
+            logger.debug("[Node lock] Released peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+
+    def _unit_requested_lock(self, unit: ops.Unit):
+        """Whether unit requested lock."""
+        assert self._relation
+        value = self._relation.data[unit].get("lock-requested")
+        if not value:
+            return False
+        value = json.loads(value)
+        if not isinstance(value, bool):
+            raise ValueError
+        return value
```

Comment:
Perhaps - safer and more concise (asserts are strongly discouraged outside of testing code - they can even be disabled when running with `-O`. Raise a specific exception if needed instead):
```suggestion
    def _unit_requested_lock(self, unit: ops.Unit) -> bool:
        """Whether unit requested lock."""
        if not self._relation:
            return False
            
        return json.loads(
            self._relation.data[unit].get("lock-requested", "false")
        )
```

Reply:
Okay - but just for the sake of clarity and completeness - `failing silently` will not happen, as what I'm advocating for is to let `self._relation.data[unit]` fail in the next line.

The other comment was with regards to the check:
```
value = json.loads(value)
if not isinstance(value, bool):
```
when would this check evaluate to true?

Reply:
I think in this case we should have a bit more predictability and trust in the code. If we do not expect unpredictable input on the `lock-requested` entry of the peer data, we should trust that and not be overly defensive here